Add backup/restore support for EEPROM settings

thehenkan opened the pull request:
This adds a button to dump EEPROM settings to a JSON file, and a button to restore them by uploading a backup.

obra commented:
I was starting to look at this today. I think my assumption had been that
it would make more sense in the import/export ui where we do keymaps. For
most users, I suspect they would expect keymaps and config to be bundled
together. (and the existing export field  should be exporting the config)

What do you think?
On Fri, Jul 11, 2025 at 10:59 PM thehenkan ***@***.***> wrote:

> *thehenkan* left a comment (keyboardio/Chrysalis#1394)
> <https://github.com/keyboardio/Chrysalis/pull/1394#issuecomment-3064726677>
>
> Ping @obra <https://github.com/obra>
>
> This is what it looks like in the settings page:
> eeprom-settings.png (view on web)
> <https://github.com/user-attachments/assets/6949d47f-3df8-45fe-af4f-207dfef4eee0>
>
> Let me know if you had other plans for this feature - I'm happy to move it
> around in the UI if you'd prefer it to be somewhere else.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Chrysalis/pull/1394#issuecomment-3064726677>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAALC2AOKEXH4AQYGRFTR2D3ICP33AVCNFSM6AAAAACBAGQLACVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTANRUG4ZDMNRXG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

thehenkan commented:
Yeah that makes sense to me. Do you mean as in the existing backup/restore buttons for the keyboard layout should just be extended to encompass all EEPROM settings instead of just the keymap, or simply moving the buttons I added to that screen?

obra commented:
So, my recollection is that the existing backup button *already* includes the eeprom settings. We just never handled them on import.  Which makes me feel like adding a checkbox for the restore button to "overwrite keyboard settings" would make more sense than new buttons. And I think that it *might* make sense to have that button checked by default.  This would mean that restore from an automated backup that gets downloaded during a firmware flash would become easy and straightforward. Which would be amazing.

thehenkan commented:
It's true that the backup button already includes the eeprom settings. I've been playing around a bit with adding that to the layout restore button just now. One thing I noticed was that the layout restore loads the layout into the UI for inspection, and then has a "Save changes" button for actually writing to the device, whereas the functionality I added to the settings page directly wrote the settings to the device and then updated the settings UI to reflect the new state.

We could make the full eeprom restore also do a separate upload & save stages, but the current layout backup screen doesn't offer a way to visualise the changes. You can't even go to the settings screen to see them without discarding the unsaved changes. Conversely, having the button sometimes write directly and sometimes not would of course be very unintuitive behaviour.

So far I've thought of merging the "My Keyboard" settings page into the layout backup page, or alternatively adding a list of settings that were changed (and their previous and new states) but without an option to change them. What do you think?

thehenkan commented:
I moved it to the existing restore button now. The behaviour of the restore can be set to only restore layout via a drop down menu. I also added a basic table showing old and new values for loaded settings, but only for the changed settings. Let me know what you think.


<img width="1048" height="1007" src="https://github.com/user-attachments/assets/f758d69f-33b1-48f9-b122-c623e0dd3851" />


<img width="434" height="171"  src="https://github.com/user-attachments/assets/761365fc-1397-4ac8-abe5-754b7d56be51" />


<img width="485" height="181"  src="https://github.com/user-attachments/assets/a714f1dc-ce25-4985-b979-2bfa9d006309" />


<img width="1054" height="1018"   src="https://github.com/user-attachments/assets/f37317a4-bbf6-4c1a-b033-3accc1a915d2" />

obra commented:
Thanks! Sorry it took a minute to get this merged
